In [41]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import TfidfVectorizer

In [2]:
df = pd.read_json('Books_small.json',lines = True)
df.head(1)

reviewerID        asin             reviewerName helpful  \
0  A1E5ZR1Z4OQJG  1495329321  Pure Jonel "Pure Jonel"  [0, 0]   

                                          reviewText  overall  \
0  Da Silva takes the divine by storm with this u...        4   

                  summary  unixReviewTime   reviewTime  
0  An amazing first novel      1396137600  03 30, 2014

In [3]:
def func(x):
    
    if x <= 2:
        return 'NEGATIVE'
    
    elif x == 3:
        
        return 'NEUTRAL'
    
    else:
        return 'POSITIVE'

In [4]:
df['overall'] = df['overall'].apply(func)

In [5]:
df['overall'].value_counts()

POSITIVE    835
NEUTRAL     103
NEGATIVE     62
Name: overall, dtype: int64

In [6]:
reviews = pd.DataFrame(df[['reviewText','overall']])

In [7]:
encoding = {"overall":     {"POSITIVE": 1, "NEGATIVE": 2 , "NEUTRAL":3},
               }

In [8]:
reviews = reviews.replace(encoding)

In [9]:
reviews['overall'].value_counts()

1    835
3    103
2     62
Name: overall, dtype: int64

In [10]:
data = reviews.values

In [11]:
X  = data[:,0]
y  = data[:,1]

In [12]:
print(X.shape)
print(y.shape)

(1000,)
(1000,)


In [13]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.33, random_state=42)
print(len(X_train))
print(len(X_test))

670
330


In [30]:
y_train = y_train.astype('int')
y_test = y_test.astype('int')

# TFIDF Vectorizer 

In [14]:
vectorizer = TfidfVectorizer()

vect_trainx = vectorizer.fit_transform(X_train)
vect_testx = vectorizer.transform(X_test)

print(vect_trainx[0].shape)

(1, 7372)


In [19]:
type(y_train[0])

int

# Classification

In [54]:
from sklearn.neighbors import KNeighborsClassifier
np.random.seed(1)

neigh = KNeighborsClassifier(n_neighbors=3)
neigh.fit(vect_trainx, y_train)

KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
                     metric_params=None, n_jobs=None, n_neighbors=3, p=2,
                     weights='uniform')

In [55]:
neigh.predict(vect_testx[0])

array([1])

In [56]:
neigh.score(vect_testx,y_test)

0.8545454545454545

In [57]:
# F1 Scores
from sklearn.metrics import f1_score
f1_score(y_test, neigh.predict(vect_testx), average=None, labels=[3,2,1])

array([0.19512195, 0.        , 0.92205638])

We get highly biased on positive review and its obvious because of data have more positive rows.
Here is ratio, positive have 835,negative have 103 and neutral have 62 rows.
1    835
3    103
2     62
In future we improve this model, IA.